nixos/modules/services/hardware/display.nix

```diff
@@ -0,0 +1,235 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.hardware.display;
+
+  edidKernelParams = lib.trivial.pipe cfg.outputs [
+    (lib.attrsets.filterAttrs (_: spec: spec.edid != null))
+    (lib.mapAttrsToList (output: spec: "${output}:edid/${spec.edid}"))
+    (builtins.concatStringsSep ",")
+    (p: lib.optional (p != "") "drm.edid_firmware=${p}")
+  ];
+in
+{
+  meta.doc = ./display.md;
+  meta.maintainers = with lib.maintainers; [
+    nazarewk
+  ];
+
+  options = {
+    hardware.display.edid.enable = lib.mkOption {
+      type = with lib.types; bool;
+      default = cfg.edid.packages != null;
+      defaultText = lib.literalExpression "config.hardware.display.edid.packages != null";
+      description = ''
+        Enables handling of EDID files
+      '';
+    };
+
+    hardware.display.edid.packages = lib.mkOption {
+      type = with lib.types; listOf package;
+      default = [ ];
+      description = ''
+        List of packages containing EDID binary files at `$out/lib/firmware/edid`.
+        Such files will be available for use in `drm.edid_firmware` kernel
+        parameter as `edid/<filename>`.
+
+        You can craft one directly here or use sibling options `linuxhw` and `modelines`.
+      '';
+      example = lib.literalExpression ''
+        [
+          (pkgs.runCommand "edid-custom" {} '''
+            mkdir -p "$out/lib/firmware/edid"
+            base64 -d > "$out/lib/firmware/edid/custom1.bin" <<'EOF'
+            <insert your base64 encoded EDID file here `base64 < /sys/class/drm/card0-.../edid`>
+            EOF
+          ''')
+        ]
+      '';
+      apply = list:
+        if list == [ ] then null else
+        (pkgs.buildEnv {
+          name = "firmware-edid";
+          paths = list;
+          pathsToLink = [ "/lib/firmware/edid" ];
+          ignoreCollisions = true;
+        }) // {
+          compressFirmware = false;
```

Comment:
We shouldn't actually need `compressFirmware = false;`, since the kernel is built with compressed firmware support by default. But it looks trivial either way: trivial space savings with compression (for edid bins) vs. added complexity in nixos config.

Reply:
it is needed, I think code paths in kernel/drivers do not decompress EDID files or at least not properly. Don't remember the details, but i tried to investigate quite extensively (without much knowledge of C/C++ itself)

Reply:
But my archlinux partition is using a zstd compressed edid firmware file without issue. Using Archlinux vanilla kernel v6.8.9.arch1-2.
```
$ cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-linux root=UUID=eb3729f4-547c-42a7-86b4-6020f9a31762 rw loglevel=3 quiet nowatchdog intel_iommu=on iommu=pt zswap.enabled=0 ibt=off amdgpu.ppfeaturemask=0xffffffff drm.edid_firmware=edid/g70nc_edid.bin

$ ls /usr/lib/firmware/edid/
g70nc_edid.bin.zst
```
I compressed it manually with `zstd -19 g70nc_edid.bin`

Reply:
@nazarewk  With 1ed5df1e703135d2d6814a761ce67e6e8073af30 reverted, my nixos partition loads zstd compressed edid firmware ( as of nixos-unstable commit ad57eef4ef0659193044870c731987a6df5cf56b , kernel v6.9.3).

```
$ file /run/current-system/firmware/edid/g70nc_edid.bin.zst
/run/current-system/firmware/edid/g70nc_edid.bin.zst: Zstandard compressed data (v0.8+), Dictionary ID: None
```

Here is the relevant nixos config:
```nix
{ config, lib, pkgs, ... }:
let
  samsung-G70NC-edid = pkgs.runCommand "g70nc_edid.bin" { } ''
    mkdir -p $out/lib/firmware/edid
    cp ${./samsung_G70NC_edid_fix.bin} $out/lib/firmware/edid/g70nc_edid.bin
  '';
in {
  boot.kernelParams = [
    "intel_iommu=on" "iommu=pt"
    "zswap.enabled=0"  # because using zram
    "amdgpu.ppfeaturemask=0xffffffff"  # unlock amdgpu voltage and clock adjustments
    "ibt=off"  # for bcachefs
    "drm.edid_firmware=edid/g70nc_edid.bin"
  ];
  hardware.firmware = [
    samsung-G70NC-edid
  ];
}
```